File: kozbookapp/app/src/main/java/example/win10/kozbookapp/ui/fragment/BookFragment.java
package example.win10.kozbookapp.ui.fragment;

import android.annotation.SuppressLint;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.core.content.res.ResourcesCompat;
import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.widget.Button;
import android.widget.GridLayout;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.TextView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.Objects;

import androidx.fragment.app.FragmentTransaction;
import example.win10.kozbookapp.R;
import example.win10.kozbookapp.model.AccessBook;
import example.win10.kozbookapp.model.Book;
import example.win10.kozbookapp.model.Library;
import example.win10.kozbookapp.model.Location;
import example.win10.kozbookapp.viewmodel.LibraryViewModel;

public class BookFragment extends Fragment implements View.OnClickListener {
    private LibraryViewModel mViewModel;
    private Book book;

    private LinearLayout bookBarLayout;
    private TextView nameText;
    private TextView authorText;
//    private ImageView bookCover;
    private TableLayout locationTable;
    private TableLayout accessTable;

    public BookFragment(LibraryViewModel mViewModel){
        super();
        this.mViewModel = mViewModel;
    }

    public static BookFragment newInstance(LibraryViewModel mViewModel) {
        return new BookFragment(mViewModel);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.book_fragment, container, false);

        this.bookBarLayout = v.findViewById(R.id.infoBarLayout);
        this.nameText = v.findViewById(R.id.bookName);
        this.authorText = v.findViewById(R.id.authorName);
//        this.bookCover = v.findViewById(R.id.coverView);

        this.locationTable = v.findViewById(R.id.locationTable);
        this.accessTable = v.findViewById(R.id.accessTable);

//        Button b = v.findViewById(R.id.backBtn);
//        b.setOnClickListener(this);

        return v;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        final ViewTreeObserver vto = this.requireView().getViewTreeObserver();
        if (vto.isAlive()) {
            vto.addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
                @Override
                public void onGlobalLayout() {
                    vto.removeOnGlobalLayoutListener(this);
                    int viewWidth = bookBarLayout.getMeasuredWidth();
                    getSelectedBook(viewWidth);
                }
            });
        }
    }

    private void getSelectedBook(int viewWidth){
        this.mViewModel.getSelectedBook().observe(getViewLifecycleOwner(), book -> {
            this.book = book;

            this.populateBookInfoBar(viewWidth);
            this.populateLocationTable();
            this.populateAccessTable();
        });
    }

    private void populateBookInfoBar(int viewWidth){
        this.nameText.setText(book.getName());
        this.authorText.setText(book.getAuthor().getName());
        this.authorText.setOnClickListener(v->{
            this.mViewModel.setSelectedAuthor(this.book.getAuthor());
            this.changeFragment(new AuthorFragment(this.mViewModel));
        });

//        this.bookCover.setLayoutParams(new LinearLayout.LayoutParams(viewWidth/2, viewWidth/2));

//        Drawable myDrawable = ResourcesCompat.getDrawable(getResources(), R.drawable.no_cover, null);
//        this.bookCover.setImageDrawable(myDrawable);
    }

    @SuppressLint("ResourceAsColor")
    private void populateLocationTable(){
        this.locationTable.removeAllViews();
        for(int i=0; i<this.book.getLocations().size();i++) {
            Location loc = this.book.getLocations().get(i);
            TableRow tr_head = this.addLocationTableRow(loc, i);

            if(i==0){
                tr_head.setBackgroundColor(R.color.latest);
            }

            this.locationTable.addView(tr_head, new TableLayout.LayoutParams(
                    TableLayout.LayoutParams.MATCH_PARENT,
                    TableLayout.LayoutParams.WRAP_CONTENT));
        }
    }
    private TableRow addLocationTableRow(Location loc, int i){
        TableRow tr_head = new TableRow(this.getContext());
        tr_head.setId(i + 1);
        tr_head.setBackgroundColor(Color.GRAY);
        tr_head.setLayoutParams(new TableLayout.LayoutParams(
                TableLayout.LayoutParams.MATCH_PARENT,
                TableLayout.LayoutParams.WRAP_CONTENT));

        TextView tv;
        tv= this.addTextView(loc.getName(), i+1);
        tr_head.addView(tv);
        tv= this.addTextView(loc.getExtra_info(), i+1);
        tr_head.addView(tv);

        return tr_head;
    }

    private void populateAccessTable(){
        this.accessTable.removeAllViews();

        mViewModel.getSelectedLibrary().observe(getViewLifecycleOwner(), p -> {
            for (int i = 0; i < p.getAccessBooks().size(); i++) {
                AccessBook ab = p.getAccessBooks().get(i);
                TableRow tr_head = this.addAccessTableRow(ab, i);

                this.accessTable.addView(tr_head, new TableLayout.LayoutParams(
                        TableLayout.LayoutParams.MATCH_PARENT,
                        TableLayout.LayoutParams.WRAP_CONTENT));
            }
        });
    }

    private TableRow addAccessTableRow(AccessBook ab, int i){
        TableRow tr_head = new TableRow(this.getContext());
        tr_head.setId(i + 1);
        tr_head.setBackgroundColor(Color.GRAY);
        tr_head.setLayoutParams(new TableLayout.LayoutParams(
                TableLayout.LayoutParams.MATCH_PARENT,
                TableLayout.LayoutParams.WRAP_CONTENT));

        TextView tv;
        tv= this.addTextView(ab.getUser().getUsername(), i+1);
        tr_head.addView(tv);
        tv= this.addTextView((String) ab.getStartDateString(), i+1);
        tr_head.addView(tv);
        if(ab.getEndDate() != null){
            tv= this.addTextView((String) ab.getEndDateString(), i+1);
        } else{
            tv= this.addTextView("", i+1);
        }
        tr_head.addView(tv);

        return tr_head;
    }

    private TextView addTextView(String data, int i){
        TextView tv = new TextView(this.getContext());
        tv.setId(i + 111);
        tv.setText(data);
        tv.setTextColor(Color.WHITE);
        tv.setPadding(5, 20, 5, 5);
        return tv;
    }

    @Override
    public void onClick(View view) {
        if (view.getId() == R.id.backBtn) {
            this.changeFragment(new BookListFragment(this.mViewModel));
        }
    }

    public void changeFragment(Fragment newFragment){
        FragmentTransaction transaction = this.requireActivity().getSupportFragmentManager().beginTransaction();

        transaction.replace(R.id.fragment_container, newFragment);
        transaction.addToBackStack(null);

        transaction.commit();
    }
}

File: kozbookapp/app/src/main/java/example/win10/kozbookapp/ui/fragment/BookFragment.kt
package example.win10.kozbookapp.ui.fragment

import android.annotation.SuppressLint
import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import android.widget.LinearLayout
import android.widget.TableLayout
import android.widget.TableRow
import android.widget.TextView
import androidx.fragment.app.Fragment
import example.win10.kozbookapp.R
import example.win10.kozbookapp.model.AccessBook
import example.win10.kozbookapp.model.Book
import example.win10.kozbookapp.model.Library
import example.win10.kozbookapp.model.Location
import example.win10.kozbookapp.viewmodel.LibraryViewModel

class BookFragment(private val mViewModel: LibraryViewModel) : Fragment(), View.OnClickListener {
    private var book: Book? = null
    private var bookBarLayout: LinearLayout? = null
    private var nameText: TextView? = null
    private var authorText: TextView? = null

    //    private ImageView bookCover;
    private var locationTable: TableLayout? = null
    private var accessTable: TableLayout? = null
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val v = inflater.inflate(R.layout.book_fragment, container, false)
        bookBarLayout = v.findViewById(R.id.infoBarLayout)
        nameText = v.findViewById(R.id.bookName)
        authorText = v.findViewById(R.id.authorName)
        //        this.bookCover = v.findViewById(R.id.coverView);
        locationTable = v.findViewById(R.id.locationTable)
        accessTable = v.findViewById(R.id.accessTable)

//        Button b = v.findViewById(R.id.backBtn);
//        b.setOnClickListener(this);
        return v
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        val vto = requireView().viewTreeObserver
        if (vto.isAlive) {
            vto.addOnGlobalLayoutListener(object : OnGlobalLayoutListener {
                override fun onGlobalLayout() {
                    vto.removeOnGlobalLayoutListener(this)
                    val viewWidth = bookBarLayout!!.measuredWidth
                    getSelectedBook(viewWidth)
                }
            })
        }
    }

    private fun getSelectedBook(viewWidth: Int) {
        mViewModel.getSelectedBook().observe(viewLifecycleOwner) { book: Book? ->
            this.book = book
            populateBookInfoBar(viewWidth)
            populateLocationTable()
            populateAccessTable()
        }
    }

    private fun populateBookInfoBar(viewWidth: Int) {
        nameText!!.text = book!!.name
        authorText!!.text = book!!.author!!.name
        authorText!!.setOnClickListener {
            mViewModel.setSelectedAuthor(book!!.author!!)
            changeFragment(AuthorFragment(mViewModel))
        }

//        this.bookCover.setLayoutParams(new LinearLayout.LayoutParams(viewWidth/2, viewWidth/2));

//        Drawable myDrawable = ResourcesCompat.getDrawable(getResources(), R.drawable.no_cover, null);
//        this.bookCover.setImageDrawable(myDrawable);
    }

    @SuppressLint("ResourceAsColor")
    private fun populateLocationTable() {
        locationTable!!.removeAllViews()
        for (i in book!!.getLocations()!!.indices) {
            val loc = book!!.getLocations()!![i]
            val trHead = addLocationTableRow(loc, i)
            if (loc == book!!.activeLocation) {
                trHead.setBackgroundColor(R.color.latest)
            }
            locationTable!!.addView(trHead, TableLayout.LayoutParams(
                    TableLayout.LayoutParams.MATCH_PARENT,
                    TableLayout.LayoutParams.WRAP_CONTENT))
        }
    }

    private fun addLocationTableRow(loc: Location, i: Int): TableRow {
        val trHead = TableRow(this.context)
        trHead.id = i + 1
        trHead.setBackgroundColor(Color.GRAY)
        trHead.layoutParams = TableLayout.LayoutParams(
                TableLayout.LayoutParams.MATCH_PARENT,
                TableLayout.LayoutParams.WRAP_CONTENT)
        var tv: TextView = addTextView(loc.name, i + 1)
        trHead.addView(tv)
        tv = addTextView(loc.extraInfo, i + 1)
        trHead.addView(tv)
        return trHead
    }

    private fun populateAccessTable() {
        accessTable!!.removeAllViews()
        mViewModel.selectedLibrary.observe(viewLifecycleOwner) { p: Library ->
            for (i in p.getAccessBooks()!!.indices) {
                val ab = p.getAccessBooks()!![i]
                val trHead = addAccessTableRow(ab, i)
                accessTable!!.addView(trHead, TableLayout.LayoutParams(
                        TableLayout.LayoutParams.MATCH_PARENT,
                        TableLayout.LayoutParams.WRAP_CONTENT))
            }
        }
    }

    private fun addAccessTableRow(ab: AccessBook, i: Int): TableRow {
        val trHead = TableRow(this.context)
        trHead.id = i + 1
        trHead.setBackgroundColor(Color.GRAY)
        trHead.layoutParams = TableLayout.LayoutParams(
                TableLayout.LayoutParams.MATCH_PARENT,
                TableLayout.LayoutParams.WRAP_CONTENT)
        var tv: TextView = addTextView(ab.user!!.username, i + 1)
        trHead.addView(tv)
        tv = addTextView(ab.startDateString as String, i + 1)
        trHead.addView(tv)
        tv = if (ab.endDate != null) {
            addTextView(ab.endDateString as String, i + 1)
        } else {
            addTextView("", i + 1)
        }
        trHead.addView(tv)
        return trHead
    }

    private fun addTextView(data: String?, i: Int): TextView {
        val tv = TextView(this.context)
        tv.id = i + 111
        tv.text = data
        tv.setTextColor(Color.WHITE)
        tv.setPadding(5, 20, 5, 5)
        return tv
    }

    override fun onClick(view: View) {
        if (view.id == R.id.backBtn) {
            changeFragment(BookListFragment(mViewModel))
        }
    }

    private fun changeFragment(newFragment: Fragment?) {
        val transaction = requireActivity().supportFragmentManager.beginTransaction()
        transaction.replace(R.id.fragment_container, newFragment!!)
        transaction.addToBackStack(null)
        transaction.commit()
    }

    companion object {
        fun newInstance(mViewModel: LibraryViewModel): BookFragment {
            return BookFragment(mViewModel)
        }
    }
}